### PR TITLE
Add schedule task condition configuration

### PR DESCRIPTION
**Pull request by lornasong**
Supports configuring a task with a schedule condition allowing users to
configure a task that monitors services in task.services but only execute
on a configured schedule rather than be dynamically triggered on changes.

Details:
 - schedule condition's cron field is required
 - uses the hashicorp fork of `cronexpr` to validate cron expression
 - disables buffer_period and logs a warning if task's buffer_period is
 configured. buffer_period does not make sense to be enabled for a scheduled
 task

Example config:
```
task {
  name = "scheduled_task"
  description = "run task every 10 seconds"
  services = ["web", "db"]
  source = "path/to/module"
  condition "schedule" {
    cron = "*/10 * * * * * *"
  }
}
```
Part of: https://github.com/hashicorp/consul-terraform-sync/issues/308

Subsequent prs
 - Consume schedule condition configuration to trigger task on schedule
 - Support configuring monitoring services by regexp
 - Support configuring monitoring kv

### CODE REVIEW

File: config/task.go [hunk 1]
```diff
@@ -226,6 +227,20 @@ func (c *TaskConfig) Finalize(bp *BufferPeriodConfig, wd string) {
 		c.TFVersion = String("")
 	}
 
+	bp := globalBp
+	if _, ok := c.Condition.(*ScheduleConditionConfig); ok {
+		// disable buffer_period for schedule condition
+		if c.BufferPeriod != nil {
+			log.Printf("[WARN] (config.task) disabling buffer_period for "+
```
Comment (lornasong) [inline config/task.go:234]: will update this log with new logging format after https://github.com/hashicorp/consul-terraform-sync/issues/377 is merged

**Comment by wilkermichael** [comment]
General question, I'm not quite following some of the short variable names, I was wondering if you could clarify the usage of a, r, and o2? Thanks!

File: config/condition_schedule_test.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+	}{
+		{
+			"empty",
+			[]string{},
+			&ScheduleConditionConfig{},
+			&ScheduleConditionConfig{
+				Cron: String(""),
+			},
+		},
+	}
```
Comment (wilkermichael) [inline config/condition_schedule_test.go:134]: For completeness, I think we should also test the other conditions, namely:
```
		{
			"empty",
			[]string{},
			&ScheduleConditionConfig{},
			&ScheduleConditionConfig{
				Cron: String(""),
			},
		},		
                 {
			"nil config",
			[]string{},
			nil,
			nil,
		},		
                {
			"chron povided",
			[]string{},
			&ScheduleConditionConfig{
				Cron: String("* * * * *"),
			},
			&ScheduleConditionConfig{
				Cron: String("* * * * *"),
			},
		},
```

**Comment by lornasong** [comment]
Hi @wilkermichael a, r, o are used in a number of places so feel free to clarify if there's a specific usage that you want to know more about. Below is how I interpret them throughout our config code and test code.

Regarding 'a' - I think this is mostly seen in tests as one "input" for whatever is being tested tested e.g. a.testFunc(). It may have an accompanying 'b' which is another "input" a.testFunc(b)

Regarding 'r' - In tests it is often used to indicate expected returned object. This is used in our Merge() code like `r := c.Copy()` in which case I interpret 'r' to mean the replica

Regarding 'o' - I typically think of this is as meaning 'other'. For example, in Merge() "o" is the other config to merge in.

File: config/condition_schedule.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+	if o2.Cron != nil {
+		r2.Cron = StringCopy(o2.Cron)
+	}
+
+	return r2
+}
+
+// Finalize ensures there no nil pointers.
+func (c *ScheduleConditionConfig) Finalize(services []string) {
```
Comment (wilkermichael) [inline config/condition_schedule.go:61]: maybe since services isn't actually used, and since we're just adding it to satisfy the interface, we can make this anonymous?

`func (c *ScheduleConditionConfig) Finalize([]string)`

**Comment by wilkermichael** [comment]
I wonder if it's worth adding a test that tests all the methods with a nil config? That way we have peace of mind that all the nil guards are in place?

**Comment by wilkermichael** [comment]
@lornasong review complete! Great work, I think being able to schedule when services trigger automation will add a lot of value

**Review by wilkermichael — APPROVED** [review]

**Comment by wilkermichael** [comment]
> Hi @wilkermichael a, r, o are used in a number of places so feel free to clarify if there's a specific usage that you want to know more about. Below is how I interpret them throughout our config code and test code.
> 
> Regarding 'a' - I think this is mostly seen in tests as one "input" for whatever is being tested tested e.g. a.testFunc(). It may have an accompanying 'b' which is another "input" a.testFunc(b)
> 
> Regarding 'r' - In tests it is often used to indicate expected returned object. This is used in our Merge() code like `r := c.Copy()` in which case I interpret 'r' to mean the replica
> 
> Regarding 'o' - I typically think of this is as meaning 'other'. For example, in Merge() "o" is the other config to merge in.

Thanks Lorna! That clarifies it

File: config/condition_schedule_test.go (hunk 2 above)
Comment (lornasong) [reply config/condition_schedule_test.go:150]: Good idea, added!

**Comment by lornasong** [comment]
Thanks for the review @wilkermichael

> I wonder if it's worth adding a test that tests all the methods with a nil config? That way we have peace of mind that all the nil guards are in place?

I just added a nil test case to Validate() in addition to your suggestion to add a nil test case to Finalize(). This will cover the nil config test case within each individual test functions.

File: config/task.go (hunk 1 above)
Comment (lornasong) [reply config/task.go]: Updated log line to:

```
logging.Global().Named(logSystemName).Named(taskSubsystemName).Warn(
	"disabling buffer_period for schedule condition. overriding "+
	"buffer_period configured for this task",
	"task_name", StringVal(c.Name), "buffer_period", c.BufferPeriod.GoString())
```
Output:
2021-09-08T12:50:12.631-0400 [WARN]  config.task: disabling buffer_period for schedule condition. overriding buffer_period configured for this task: task_name=periodic10_task buffer_period="&BufferPeriodConfig{Enabled:true, Min:8s, Max:0s}"
